alchemist/alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/actions/AvoidFlowField.kt

```diff
@@ -1,51 +1,56 @@
 package it.unibo.alchemist.model.implementations.actions
 
 import it.unibo.alchemist.model.implementations.layers.BidimensionalGaussianLayer
-import it.unibo.alchemist.model.implementations.utils.origin
+import it.unibo.alchemist.model.interfaces.Action
+import it.unibo.alchemist.model.interfaces.Environment
+import it.unibo.alchemist.model.interfaces.Layer
 import it.unibo.alchemist.model.interfaces.Molecule
+import it.unibo.alchemist.model.interfaces.Node
 import it.unibo.alchemist.model.interfaces.Pedestrian2D
+import it.unibo.alchemist.model.interfaces.Position2D
 import it.unibo.alchemist.model.interfaces.Reaction
-import it.unibo.alchemist.model.interfaces.environments.Physics2DEnvironment
-import it.unibo.alchemist.model.interfaces.movestrategies.TargetSelectionStrategy
+import it.unibo.alchemist.model.interfaces.geometry.Vector2D
 
 /**
  * Move the pedestrian towards positions of the environment with a low concentration of the target molecule.
  *
  * @param env
  *          the environment inside which the pedestrian moves.
+ * @param reaction
+ *          the reaction which executes this action.
  * @param pedestrian
  *          the owner of this action.
  * @param targetMolecule
  *          the {@link Molecule} you want to know the concentration in the different positions of the environment.
  */
-open class AvoidFlowField(
-    env: Physics2DEnvironment<Number>,
+open class AvoidFlowField<P>(
+    env: Environment<Number, P>,
     reaction: Reaction<Number>,
     pedestrian: Pedestrian2D<Number>,
     targetMolecule: Molecule
-) : FlowFieldSteeringAction<Number>(
-    env,
-    reaction,
-    pedestrian,
-    targetMolecule,
-    { molecule ->
-        val currentPosition = env.getPosition(pedestrian)
-        val layer = env.getLayer(molecule).get()
-        val currentConcentration = layer.getValue(currentPosition).toDouble()
-        this.map { it to layer.getValue(it).toDouble() }
-            .filter { it.second < currentConcentration }
-            .minBy { it.second }?.first ?: currentPosition
-    },
-    TargetSelectionStrategy {
-        with(env.getLayer(targetMolecule)) {
-            val p = env.getPosition(pedestrian)
-            if (p == null || isEmpty || get() !is BidimensionalGaussianLayer<*>) {
-                env.origin()
-            } else {
-                with(get() as BidimensionalGaussianLayer) {
-                    p + (p - env.makePosition(centerX, centerY))
-                }
+) : FlowFieldSteeringAction<P>(env, reaction, pedestrian, targetMolecule)
+    where
+        P : Position2D<P>,
+        P : Vector2D<P> {
+
+    override fun cloneAction(n: Node<Number>, r: Reaction<Number>): Action<Number> =
+        AvoidFlowField(env, r, n as Pedestrian2D<Number>, targetMolecule)
+
+    override fun List<P>.selectPosition(layer: Layer<Number, P>, currentConcentration: Double): P = this
+        .toMutableList()
```

Comment:
Considering that you modify the collection multiple times, have you considered using `Sequence` instead?

Reply:
Moved to `Sequence`, I didn't remember they were faster.